app/services/user-agent-service/datomstorage-sqlite.js

```diff
@@ -0,0 +1,161 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+
+import {
+  datascript,
+  mori,
+  helpers,
+} from 'datascript-mori';
+
+import * as datomstore from './datomstore';
+import { DatomStorageSchemaV1 } from './datomstorage-schema';
+
+const { vector } = mori;
+const { DB_ADD, DB_RETRACT } = helpers;
+
+const d = datascript.core;
+const djs = datascript.js;
+
+export class DatomStorageSqlite {
+  constructor(db) {
+    this.db = db;
+  }
+
+  async init(sqlSchema = new DatomStorageSchemaV1()) {
+    await this.db.exec('PRAGMA foreign_keys = ON');
+    const v = await sqlSchema.createOrUpdate(this);
+
+    if (v !== sqlSchema.version) {
+      throw new Error(`Incorrect version ${v}; expected ${sqlSchema.version}.`);
+    }
+  }
+
+  async close() {
+    this.db.close();
+    return true;
+  }
+
+  async userVersion() {
+    return await this.db
+      .get('PRAGMA user_version')
+      .then((row) => Promise.resolve(row.user_version));
+  }
+
+  async listen(report) {
+    const txData = mori.get(report, mori.keyword('tx-data'));
+
+    await this.db.inTransaction(() => {
+      // Write 150 5-tuples at a time.  SQLite has a limit of 999 bound variables per statement.
+      let ps = mori.map((txChunk) => {
+        const qs = `(?, ?, ?, ?, ?)${', (?, ?, ?, ?, ?)'.repeat(mori.count(txChunk) - 1)}`;
+        return this.db
+          .run(`INSERT INTO transactions (e, a, v, tx, added) VALUES ${qs}`,
+               mori.intoArray(mori.mapcat((tx) => mori.vector(tx.e,
+                                                              tx.a,
+                                                              tx.v,
+                                                              tx.tx,
+                                                              tx.added ? 1 : 0), txChunk)));
+      }, mori.partitionAll(150, txData));
+
+      // ps is a lazy sequence.  intoArray forces, like `seq`.
+      ps = mori.intoArray(ps);
+      return Promise.all(mori.intoArray(ps));
+    });
+  }
+
+  async getLastTxInSnapshot() {
+    const [{ lastTx }] =
+      await this.db.all('SELECT COALESCE(MAX(tx), -1) AS lastTx FROM datoms');
```

Comment:
I don't believe there's an index that will apply to make this efficient, so this will be a full table walk.

You might consider putting the current max into a separate table, or adding an index, if this degrades open perf.